jaxopt/_src/nonlinear_cg.py

```diff
@@ -0,0 +1,195 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Nonlinear conjugate gradient algorithm"""
+
+from typing import Any
+from typing import Callable
+from typing import NamedTuple
+from typing import Optional
+
+from dataclasses import dataclass
+
+import jax
+import jax.numpy as jnp
+
+from jaxopt._src import base
+from jaxopt._src.backtracking_linesearch import BacktrackingLineSearch
+from jaxopt.tree_util import tree_vdot
+from jaxopt.tree_util import tree_scalar_mul
+from jaxopt.tree_util import tree_add_scalar_mul
+from jaxopt.tree_util import tree_sub
+from jaxopt.tree_util import tree_div
+from jaxopt.tree_util import tree_l2_norm
+
+
+class NonlinearCGState(NamedTuple):
+  """Named tuple containing state information."""
+  iter_num: int
+  stepsize: float
+  error: float
+  value: float
+  grad: any
+  p: jnp.ndarray
+  aux: Optional[Any] = None
+
+
+@dataclass(eq=False)
+class NonlinearCG(base.IterativeSolver):
+  """
+    Nonlinear Conjugate Gradient Solver.
+  Attributes:
+    fun: a smooth function of the form ``fun(x, *args, **kwargs)``.
+    method: which variant to calculate the beta parameter in Nonlinear CG.
+    ``{"polak-ribiere", "fletcher-reeves", "hestenes-stiefel"}``
```

Comment:
```suggestion
    [two spaces]Possible values are `"polak-ribiere"`, `"fletcher-reeves"` and "`hestenes-stiefel"`.
```

Reply:
Make sure you add the two spaces at the beginning of the line, otherwise it may not be considered as the attribute description.

Reply:
I did not know that. thanks to let me know. It looks fine to me now.